dash-renderer/src/persistence.js

```diff
@@ -0,0 +1,346 @@
+ *   to enable persistence, omit or set `false` to disable. For more complex
+ *   scenarios, use any truthy value, and change to a *different* truthy value
+ *   when you want the persisted values cleared. (modeled off `uirevision` in)
+ *   plotly.js
+ *   Typically should have no default, but the other persistence props should
+ *   have defaults, so all a user needs to do to enable persistence is set this
+ *   one prop.
+ *
+ * - `persisted_props`: array of prop names allowed to persist. Normally should
+ *   default to the full list of supported props, so they can all be enabled at
+ *   once. The main exception to this is if there's a prop that *can* be
+ *   persisted but most users wouldn't want this.
+ *
+ * - `persistence_type`: one of "local", "session", or "memory", just like
+ *   `dcc.Store`. But the default here should be "local" because the main use
+ *   case is to maintain settings across reloads.
+ *
+ * In addition, if any props require special behavior - particularly for storing
+ * a piece nested inside a larger prop, a component can define a class property
+ * (not a React prop) `persistenceTransforms`. This should be an object:
+ * {
+ *   [propName]: {
+ *     extract: propValue => valueToStore,
+ *     apply: (storedValue, propValue) => newPropValue
+ *   }
+ * }
+ * - `extract` turns a prop value into a reduced value to store.
+ * - `apply` puts an extracted value back into the prop.
+ * You only need to define these for the props that need them.
+ * It's important that `extract` pulls out *only* the relevant pieces of the
+ * prop, because persistence is only maintained if the extracted value of the
+ * prop before applying persistence is the same as it was before the user's
+ * changes.
+ */
+
+import {
+    difference,
+    equals,
+    filter,
+    forEach,
+    forEachObjIndexed,
+    keys,
+    lensPath,
+    set,
+    type,
+} from 'ramda';
+
+import Registry from './registry';
+
+const storePrefix = '_dash_persistence.';
+const UNDEFINED = 'U';
+
+/*
+ * Does a key fit this prefix? Must either be an exact match
+ * or a scoped match - exact prefix followed by a dot (then anything else)
+ */
+function keyPrefixMatch(prefix) {
+    return key =>
+        key === prefix || key.substr(0, prefix.length + 1) === prefix + '.';
+}
+
+class WebStore {
+    constructor(storage) {
+        this._storage = storage;
+    }
+
+    hasItem(key) {
+        return this._storage.getItem(storePrefix + key) !== null;
+    }
+
+    getItem(key) {
+        // note: _storage.getItem returns null on missing keys
+        // and JSON.parse(null) returns null as well
+        const gotVal = this._storage.getItem(storePrefix + key);
+        return gotVal === UNDEFINED ? void 0 : JSON.parse(gotVal);
+    }
+
+    setItem(key, value) {
+        const setVal = value === void 0 ? UNDEFINED : JSON.stringify(value);
+        this._storage.setItem(storePrefix + key, setVal);
+    }
```

Comment:
I think `setItem` calls might need to be more defensive.

1. Some browsers and some setups will not define `window.localStorage` and/or `window.sessionStorage` (I do not think this is a concern in 2019 anymore... -- everything >= IE8 supports this)
2. Some browsers will expose local/session storage with storage size of 0 byte - user configurations on certain browsers could also cause an issue here
3. `setItem` willthrow an error if the storage space is exceeded - this will disrupt the app's render flow

e.g. Each prop in the loop at https://github.com/plotly/dash/pull/903/files#diff-1f9fdd7834ce2c4e31ca7ce6aa890a1fR203 could be guarded in a `try ... catch` and/or LocalStorage be defaulted to MemoryStorage

More generally, should failure to store be conveyed to the user in debug mode and how?

Reply:
Right, good call. (1) I think you're right, not an issue. But (2) and (3) are worth protecting against. Also I notice `dcc.Store` needs this treatment -> https://github.com/plotly/dash-core-components/issues/635

(2) seems like all we can do is fall back on memory.
(3) we perhaps have some more options: I can imagine in fact app developers filling up their storage with data from old apps - perhaps we can clear out data associated with IDs that don't exist on the page? Certainly not ideal, as those IDs may pop up later. I suppose we could do better by including a timestamp with the data, and delete the oldest ones first. That's a lot of overhead though for a hopefully rare edge case.

Reply:
> (2) seems like all we can do is fall back on memory.

Yes. At least in memory will preserve the behavior while in a session.

> (3) we perhaps have some more options: I can imagine in fact app developers filling up their storage with data from old apps `[...]`

Interesting. All callbacks are defined though.. maybe we could use (dependencyGraph + persisted_props) to identify ids that are not significant for this app? Seems weird..

Reply:
Defensive behavior added in https://github.com/plotly/dash/pull/903/commits/83b8cd3cb0910dd79e00f0cc2d1fdbfa25d7768e

- Don't try to access stores until requested; That way the app is ready, so we can `dispatch` any errors to devtools.
- Try to add some reasonably large data to the store - I chose 65K chars (130kB) as a meaningful quantity (more than most persistence needs, unless you store table `data` or something) but still far less than normal webstore limits
- If it fails on initialization, try to clear out the persistence-related keys and try again
- Then give up and use memory if it still fails
- If it fails later on (not during init), don't try to recover or fall back - just report the error to both the console and devtools. We could explore ways to be smarter about this but it's more involved - see TODO https://github.com/plotly/dash/pull/903/commits/83b8cd3cb0910dd79e00f0cc2d1fdbfa25d7768e#diff-1f9fdd7834ce2c4e31ca7ce6aa890a1fR134